File: examples/python/cov_subscriptions.py
"""Change of Value (COV) subscription example.

Demonstrates:
- Subscribing to COV notifications on an analog input
- Receiving real-time value changes via async iterator
- Unsubscribing when done
"""

import asyncio

from rusty_bacnet import (
    BACnetClient,
    BACnetServer,
    ObjectIdentifier,
    ObjectType,
    PropertyIdentifier,
    PropertyValue,
)


async def main():
    # Start a server
    server = BACnetServer(
        device_instance=1000, device_name="Sensor Controller", port=0
    )
    server.add_analog_input(instance=1, name="Zone Temp", units=62, present_value=72.0)
    await server.start()
    addr = await server.local_address()
    print(f"Server at {addr}")

    ai1 = ObjectIdentifier(ObjectType.ANALOG_INPUT, 1)

    async with BACnetClient(port=0) as client:
        # Subscribe to COV (confirmed, 60 second lifetime)
        await client.subscribe_cov(
            addr,
            subscriber_process_identifier=1,
            monitored_object_identifier=ai1,
            confirmed=True,
            lifetime=60,
        )
        print("Subscribed to COV on AI:1")

        # Start a listener task
        async def listen_for_changes():
            async for notif in client.cov_notifications():
                obj = notif.monitored_object_identifier
                print(f"\n  COV from {obj}:")
                for v in notif.values:
                    pid = v["property_id"]
                    val = v["value"]
                    print(f"    {pid}: {val.value if val else 'N/A'}")

        listener = asyncio.create_task(listen_for_changes())

        # Simulate value changes on the server
        for temp in [73.0, 74.5, 71.0, 75.0]:
            await asyncio.sleep(0.3)
            await server.write_property_local(
                ai1,
                PropertyIdentifier.PRESENT_VALUE,
                PropertyValue.real(temp),
            )
            print(f"Server wrote: {temp}")

        await asyncio.sleep(0.5)

        # Unsubscribe
        await client.unsubscribe_cov(
            addr,
            subscriber_process_identifier=1,
            monitored_object_identifier=ai1,
        )
        print("\nUnsubscribed from COV")
        listener.cancel()

    await server.stop()


if __name__ == "__main__":
    asyncio.run(main())

File: examples/kotlin/CovSubscriptions.kt
/**
 * Change of Value (COV) subscription example.
 *
 * Demonstrates:
 * - Subscribing to COV notifications on an analog input
 * - Receiving real-time value changes via coroutine
 * - Server-side writes triggering COV notifications
 * - Unsubscribing when done
 */

import kotlinx.coroutines.*
import uniffi.bacnet_java.*

fun main() = runBlocking {
    // --- Server ---
    val serverPort: UShort = 47871u
    val server = BacnetServer(
        deviceInstance = 1000u,
        deviceName = "Sensor Controller",
        config = TransportConfig.Bip(
            address = "127.0.0.1",
            port = serverPort,
            broadcastAddress = "127.0.0.255",
        ),
    )
    server.addAnalogInput(1u, "Zone Temp", 62u, 72.0f)
    server.start()

    val serverAddr = "127.0.0.1:$serverPort"
    println("Server at $serverAddr")

    // --- Client ---
    val client = BacnetClient(
        config = TransportConfig.Bip(
            address = "127.0.0.1",
            port = 0u,
            broadcastAddress = "127.0.0.255",
        ),
        apduTimeoutMs = 6000u,
    )
    client.connect()

    try {
        val analogInput: UInt = 0u
        val presentValue: UInt = 85u

        // Subscribe to COV (confirmed, 60-second lifetime)
        client.subscribeCov(
            serverAddr,
            subscriberProcessIdentifier = 1u,
            objectType = analogInput,
            objectInstance = 1u,
            confirmed = true,
            lifetime = 60u,
        )
        println("Subscribed to COV on AnalogInput:1")

        // Start a listener coroutine
        val stream = client.covNotifications()
        val listener = launch {
            while (isActive) {
                val notif = stream.next() ?: break
                println("\n  COV notification:")
                println("    device=${notif.deviceInstance} object=AI:${notif.objectInstance}")
                for (v in notif.values) {
                    println("    property ${v.propertyId}: ${v.value}")
                }
            }
        }

        // Simulate value changes on the server
        for (temp in listOf(73.0f, 74.5f, 71.0f, 75.0f)) {
            delay(300)
            server.writePropertyLocal(
                analogInput, 1u,
                presentValue,
                BacnetPropertyValue.Real(temp),
            )
            println("Server wrote: $temp")
        }

        delay(500)

        // Unsubscribe
        client.unsubscribeCov(
            serverAddr,
            subscriberProcessIdentifier = 1u,
            objectType = analogInput,
            objectInstance = 1u,
        )
        println("\nUnsubscribed from COV")
        listener.cancel()

    } finally {
        client.stop()
        server.stop()
    }
}
